File: bindings/bun/demo/03_outliers.ts
/**
 * Copyright (c) 2026 Edison Manrique Chocce
 * Licencia Dual: AGPL-3.0 | Comercial según LICENSE.
 *
 * demo/03_outliers.ts — Detección Robusta de Outliers (Hampel)
 */
import { Outliers } from "../src/index";

console.log("=".repeat(60));
console.log("  ERMC | Demo 03: Detección Robusta de Outliers (Hampel)");
console.log("=".repeat(60));

// [1] Señal sinusoidal con spikes cósmicos
const n = 20;
const signal = Array.from({ length: n }, (_, k) => Math.sin((2 * Math.PI * k) / n));
signal[5] = 1e100;
signal[13] = -1e100;
const res1 = Outliers.detectHampel(signal, 3.0);
console.log("\n[1] Señal sinusoidal con spikes cósmicos:");
console.log(`  Outliers detectados: ${res1.outliersCount} de ${n}`);
console.log(`  Posiciones: [${res1.isOutlier.map((b, i) => b ? i : -1).filter(i => i >= 0).join(", ")}]`);
console.log(`  Datos limpios (pos 4-7): [${Array.from(res1.cleanData).slice(4, 8).map(x => x.toFixed(4)).join(", ")}]`);

// [2] Temperatura con sensores defectuosos
const temps = [22.1, 22.5, 23.0, 22.8, 999.0, 22.9, 23.1, -200.0, 22.7, 23.2,
               22.6, 23.0, 22.4, 22.8, 23.1, 22.3, 22.9, 23.3, 22.5, 22.7];
const res2 = Outliers.detectHampel(temps, 2.5);
const cleanMean = Array.from(res2.cleanData).reduce((a, b) => a + b, 0) / temps.length;
console.log("\n[2] Temperatura diaria con sensores defectuosos:");
console.log(`  Outliers detectados: ${res2.outliersCount}`);
console.log(`  Posiciones corruptas: [${res2.isOutlier.map((b, i) => b ? i : -1).filter(i => i >= 0).join(", ")}]`);
console.log(`  Media limpia: ${cleanMean.toFixed(2)} °C`);

// [3] ECG con artefactos
const ecg = Array.from({ length: 50 }, (_, k) =>
  Math.sin((2 * Math.PI * k) / 8) * Math.exp(-0.05 * k)
);
ecg[10] = 50.0;
ecg[25] = -40.0;
ecg[40] = 30.0;
const res3 = Outliers.detectHampel(ecg, 3.0);
const maxBefore = Math.max(...ecg.map(Math.abs));
const meanBefore = ecg.reduce((a, b) => a + Math.abs(b), 0) / ecg.length;
const maxAfter = Math.max(...Array.from(res3.cleanData).map(Math.abs));
const meanAfter = Array.from(res3.cleanData).reduce((a, b) => a + Math.abs(b), 0) / ecg.length;
console.log("\n[3] ECG con artefactos de movimiento:");
console.log(`  Artefactos detectados: ${res3.outliersCount} de 50`);
console.log(`  Posiciones: [${res3.isOutlier.map((b, i) => b ? i : -1).filter(i => i >= 0).join(", ")}]`);
console.log(`  Peak-to-mean ANTES:   ${(maxBefore / meanBefore).toFixed(2)}x`);
console.log(`  Peak-to-mean DESPUÉS: ${(maxAfter / meanAfter).toFixed(2)}x`);

console.log("\n[OK] Demo 03 completado.\n");

File: bindings/python/demo/03_outliers.py
# Copyright (c) 2026 Edison Manrique Chocce
# Todos los derechos reservados.
# Licencia Dual: AGPL-3.0 | Comercial según LICENSE.
"""
demo/03_outliers.py — Detección Robusta de Outliers (Hampel)
Aplica el identificador de Hampel a señales reales con perturbaciones.
"""
import sys
if sys.stdout.encoding != "utf-8":
    sys.stdout.reconfigure(encoding="utf-8")

import math
sys.path.insert(0, "..")
from ermc import Outliers

print("=" * 60)
print("  ERMC | Demo 03: Deteccion Robusta de Outliers (Hampel)")
print("=" * 60)

# --- Experimento 1: Señal sinusoidal con spikes cósmicos ---
print("\n[1] Senal sinusoidal con spikes cosmicos (1e100):")
n = 20
signal = [math.sin(2 * math.pi * k / n) for k in range(n)]
signal[5] = 1e100    # spike positivo
signal[13] = -1e100  # spike negativo
res = Outliers.detect_hampel(signal, 3.0)
print(f"  Outliers detectados: {res.outliers_count} de {n}")
print(f"  Posiciones: {[i for i, b in enumerate(res.is_outlier) if b]}")
print(f"  Datos limpios (pos 4-7): {[f'{x:.4f}' for x in res.clean_data[4:8]]}")

# --- Experimento 2: Señal de temperatura con mediciones corruptas ---
print("\n[2] Temperatura diaria (grados C) con sensores defectuosos:")
temps = [22.1, 22.5, 23.0, 22.8, 999.0, 22.9, 23.1, -200.0, 22.7, 23.2,
         22.6, 23.0, 22.4, 22.8, 23.1, 22.3, 22.9, 23.3, 22.5, 22.7]
res2 = Outliers.detect_hampel(temps, 2.5)
print(f"  Outliers detectados: {res2.outliers_count}")
print(f"  Posiciones corruptas: {[i for i, b in enumerate(res2.is_outlier) if b]}")
print(f"  Media limpia: {sum(res2.clean_data) / len(res2.clean_data):.2f} C")

# --- Experimento 3: ECG sintético con artefactos de movimiento ---
print("\n[3] ECG sintetico (50 muestras) con 3 artefactos:")
ecg = [math.sin(2 * math.pi * k / 8) * math.exp(-0.05 * k) for k in range(50)]
ecg[10] = 50.0   # artefacto muscular
ecg[25] = -40.0  # movimiento del electrodo
ecg[40] = 30.0   # interferencia eléctrica
res3 = Outliers.detect_hampel(ecg, 3.0)
print(f"  Artefactos detectados: {res3.outliers_count} de 50")
print(f"  Posiciones: {[i for i, b in enumerate(res3.is_outlier) if b]}")
snr_before = max(abs(x) for x in ecg) / (sum(abs(x) for x in ecg) / len(ecg))
snr_after = max(abs(x) for x in res3.clean_data) / (sum(abs(x) for x in res3.clean_data) / len(res3.clean_data))
print(f"  Peak-to-mean ANTES:  {snr_before:.2f}x")
print(f"  Peak-to-mean DESPUES:{snr_after:.2f}x  (reducido = señal mas limpia)")

print("\n[OK] Demo 03 completado.\n")
